grpcclient: return proper nil reference from grpcclient

jsternberg opened the pull request:
The grpcclient would take an empty reference and convert it into a `(*reference)(nil)` and then store that in a `client.Reference`. This caused the `nil` check in `convertRef` to return false because it wasn't `nil` but then it panicked because it was `nil`.

`newReference` has been minorly refactored to not return an error. The method is not exported, the error value is not used, and it obscured the functionality which made it harder to use correctly.

Fixes https://github.com/moby/buildkit/issues/5379.

tonistiigi reviewed (approved):
@jsternberg Did you find out if this was a regression from some change?

jsternberg commented:
I don't know if it's reproducible there because I haven't tried, but I was able to find code that would cause the incorrect behavior as far back as when this feature was first introduced in 2018.

https://github.com/moby/buildkit/commit/303b5da7131cd11fb1934784fef61fb1a5c2a6a8#diff-7cafce0d1cc08d47c233e73305faeccde66a98cd2b952b71d1c06df4de54ce58R229-R237

It's not as obvious in that snippet, but this part:

```go
ref := &reference{id: v, c: c}
if v == "" {
	ref = nil
}
res.AddRef(k, ref)
```

The `ref` is a `*reference` because that's what type inference will give it and the `nil` is a nil `*reference`. `AddRef` takes an interface which would trigger the bug. This code got refactored at some point but it kept the same problem.

tonistiigi commented:
Can we add an integration test for this case as well.

jsternberg commented:
Added an integration test and verified that it fails without the fix. I also found another code path that had the same problem in the `RefsDeprecated` branch and fixed that one up too.
